Check if another player has the same serial

Nico8340 opened the pull request:
This pull request adds a check when connecting to ensure that players can't use the same serial, thus preventing abuses related to serial manipulation, such as for example, when they try to spoof a server's moderator's serial as their own, while they're connected to the server.

The change involved the addition of a new parameter called `check_duplicate_serials` which allows turning off the feature for development and testing servers to connect with virtual machines.

The new code is: `CD50`
The new description is: Serial already in use

Commit:
New util

New util function that allows to search for players by serial

Commit:
New enum & logic

New enum called SERIAL_DUPLICATE and logic that checks if there are no other players on the server having the same serial

Fernando-A-Rocha reviewed (approved):
Good idea.

TracerDS commented on Server/mods/deathmatch/logic/CPlayerManager.cpp:
```diff
@@ -138,6 +138,19 @@ CPlayer* CPlayerManager::Get(const char* szNick, bool bCaseSensitive)
 
+CPlayer* CPlayerManager::GetBySerial(const SString& serial) noexcept
+{
+    for (auto& player : m_Players)
```
const?

Commit:
Make GetBySerial a constant

TracerDS commented on Server/mods/deathmatch/logic/CPlayerManager.cpp:
```diff
@@ -138,6 +138,19 @@ CPlayer* CPlayerManager::Get(const char* szNick, bool bCaseSensitive)
 
+CPlayer* CPlayerManager::GetBySerial(const SString& serial) const noexcept
+{
+    for (auto& player : m_Players)
```
const?

TracerDS replied:
i.e. const auto

also std::string if possible instead of SString

TracerDS commented on Server/mods/deathmatch/logic/CGame.cpp:
```diff
@@ -1798,6 +1798,19 @@ void CGame::Packet_PlayerJoinData(CPlayerJoinDataPacket& Packet)
+            // Tell the player the problem
+            DisconnectPlayer(this, *pPlayer, CPlayerDisconnectedPacket::SERIAL_DUPLICATE);
+            return;
+        }
```
it would be better to have a const CPlayer* in an if statement:
```cpp
if (m_pPlayerManager->GetBySerial(strSerial))
{
    // Tell the console
    CLogger::LogPrintf("CONNECT: %s failed to connect (Serial already in use) (%s)\n", szNick, strIPAndSerial.c_str());

    // Tell the player the problem
    DisconnectPlayer(this, *pPlayer, CPlayerDisconnectedPacket::SERIAL_DUPLICATE);
    return;
}

// ...
const CPlayer* CPlayerManager::GetBySerial(const SString& serial) const noexcept
```

TracerDS replied:
you dont need additional variable in CGame.cpp.
Just check it in the if statement

As for the GetBySerial function, you can return a const pointer to CPlayer. I mean, both would be fine

Xenius97 commented:
This solution is not perfect because if you run multiple MTA clients on one machine (VM), you get the same serial, making localhost testing impossible.

Xenius97 commented:
Maybe you could add something like `check_double_serials` to server settings enabled by default.

ghost commented:
> This solution is not perfect because if you run multiple MTA clients on one machine (VM), you get the same serial, making localhost testing impossible.

I forgot about that, thanks for reminding.
If anyone has any ideas, feel free to share them.

ghost commented:
> Maybe you could add something like `check_double_serials` to server settings enabled by default.

Good idea, I look forward to other people's opinions as well

Fernando-A-Rocha commented:
> This solution is not perfect because if you run multiple MTA clients on one machine (VM), you get the same serial, making localhost testing impossible.

I forgot you can use VMs to run multiple MTA clients. You need 1 VM (os) per client right?

Commit:
Make suggested changes

ghost commented:
> > This solution is not perfect because if you run multiple MTA clients on one machine (VM), you get the same serial, making localhost testing impossible.
> 
> I forgot you can use VMs to run multiple MTA clients. You need 1 VM (os) per client right?

That's right

Fernando-A-Rocha commented:
> Maybe you could add something like `check_double_serials` to server settings enabled by default.

Maybe we can have a setting like `allow_unverified_clients` that is a more generic name

Commit:
Make a new parameter

Make a new parameter called check_duplicate_serials which is enabled by default

G-Moris commented:
Would this be optimized if 'std::unordered_map' and 'find' were used instead of a loop?

Fernando-A-Rocha commented on Server/mods/deathmatch/local.conf:
```diff
@@ -274,6 +274,12 @@
          Values: 0 - Off, 1 - Enabled.  Default - 0 -->
     <elementdata_whitelisted>0</elementdata_whitelisted>
 
+    <!-- This parameter determines that the server checks when the players connect that there are no other players with the same serial number. 
```
I think you also need to add this to editor.conf

It's really annoying how everything is repeated in mtaserver.conf, mtaserver.conf.template, local.conf and editor.conf. We should refactor this in a future PR.

Commit:
Merge branch 'master' into serial

Commit:
Make suggested changes

ghost commented:
> Would this be optimized if 'std::unordered_map' and 'find' were used instead of a loop?

In fact, in this use case there are no significant differences in terms of average server sizes, but for a larger search operation it is really better to use unordered_map and its method. If someone wants to refactor it, they have to modify the entire class, and this pull request is not intended for that.

Commit:
Merge branch 'master' into serial

TheNormalnij commented on Client/mods/deathmatch/logic/CPacketHandler.h:
```diff
@@ -38,6 +38,7 @@ class CPacketHandler
         ELEMENT_FAILURE,
         GENERAL_REFUSED,
         SERIAL_VERIFICATION,
+        SERIAL_DUPLICATE,
```
Move SERIAL_DUPLICATE to the end of the list. It breaks packets from an old sever

TheNormalnij commented on Server/mods/deathmatch/logic/packets/CPlayerDisconnectedPacket.h:
```diff
@@ -35,6 +35,7 @@ class CPlayerDisconnectedPacket final : public CPacket
         ELEMENT_FAILURE,
         GENERAL_REFUSED,
         SERIAL_VERIFICATION,
+        SERIAL_DUPLICATE,
```
Move SERIAL_DUPLICATE to the end of the list.

TheNormalnij commented on Server/mods/deathmatch/logic/CPlayerManager.h:
```diff
@@ -40,6 +41,7 @@ class CPlayerManager
 
     CPlayer* Get(const NetServerPlayerID& PlayerSocket);
     CPlayer* Get(const char* szNick, bool bCaseSensitive = false);
+    const CPlayer* GetBySerial(const std::string serial) const noexcept;
```
`const CPlayer*` is unnecessary. Use `CPlayer*`

TracerDS replied:
An overload would be better here instead of removing constness.

TheNormalnij replied:
`GetBySerial` is ok here. An overload is evil here.

TracerDS replied:
Why is it evil in your opinion?

TheNormalnij replied:
Because `Get` searches by nickname for a string argument currently. Two different behavior for different string types are evil

TracerDS replied:
huh? What are you talking about?

It should return const CPlayer* in const context and normal CPlayer* in non-const context. This is not evil. In fact, its standard practice that you dont seem to use resulting in making everything non-const because everything is non-const. **That** is evil

TheNormalnij replied:
> huh? What are you talking about?

About your purpose to use overload.

> It should return const CPlayer* in const context and normal CPlayer* in non-const context. This is not evil. In fact, its standard practice that you dont seem to use resulting in making everything non-const because everything is non-const. **That** is evil

This method returns `const CPlayer*` for any context. It request a const_cast or an additional method to modify returned value here. This method will work in const context too. `const` is redundant here.

TracerDS replied:
> This method returns `const CPlayer*` for any context. It request a const_cast or an additional method to modify returned value here. This method will work in const context too. `const` is redundant here.

You dont use const_cast because you are not supposed to modify the object. **Thats the point!**
If you want to modify the object, you dont use const. **Thats the point!**
Its good practice to mark it as const if you dont intent to modify whatever it returns (for pointers that is)

TheNormalnij replied:
There is no reason why `GetBySerial` should return a non-modifiable object. `GetBySerial` is const itself and it's enough.

> If you want to modify the object, you dont use const. **Thats the point!**

So you purpose an additional method `GetBySerialNonConst` to return a modifiable object.

TracerDS replied:
> GetBySerial is const itself and it's enough.

Thats not how it works.

> So you purpose an additional method GetBySerialNonConst to return a modifiable object.

No. Adding an additional overload for non-const context would be better.

TheNormalnij commented on Server/mods/deathmatch/logic/CPlayerManager.h:
```diff
@@ -17,6 +17,7 @@ class CPlayerManager;
 #include "packets/CPacket.h"
 #include "CPlayer.h"
 #include "../Config.h"
+#include <string>
```
it looks like `std::string` is defined already

TheNormalnij commented on Server/mods/deathmatch/logic/CGame.cpp:
```diff
@@ -1798,6 +1798,17 @@ void CGame::Packet_PlayerJoinData(CPlayerJoinDataPacket& Packet)
+            CLogger::LogPrintf("CONNECT: %s failed to connect (Serial already in use) (%s)\n", szNick, strIPAndSerial.c_str());
+
+            // Tell the player the problem
+            DisconnectPlayer(this, *pPlayer, CPlayerDisconnectedPacket::SERIAL_DUPLICATE);
```
Add bistream version check for old clients and use `KICK` for them

TheNormalnij commented on Server/mods/deathmatch/logic/CPlayerManager.h:
```diff
@@ -40,6 +41,7 @@ class CPlayerManager
 
     CPlayer* Get(const NetServerPlayerID& PlayerSocket);
     CPlayer* Get(const char* szNick, bool bCaseSensitive = false);
+    const CPlayer* GetBySerial(const std::string serial) const noexcept;
```
Use string_view

lopezloo commented:
> when they try to spoof a server's moderator's serial as their own

How would they know server's moderator serial?

ghost commented:
> How would they know server's moderator serial?

From servers with a loophole or a data leak

Commit:
Make suggested changes

Fernando-A-Rocha commented on Server/mods/deathmatch/logic/CPlayerManager.cpp:
```diff
@@ -138,6 +138,19 @@ CPlayer* CPlayerManager::Get(const char* szNick, bool bCaseSensitive)
+        if (player->GetSerial() == serial)
+        {
+            return player;
+        }
```
You can remove these { }

Fernando-A-Rocha replied:
```suggestion
        if (player->GetSerial() == serial)
            return player;
```

Fernando-A-Rocha reviewed (changes requested):
Other than that, good job!

Commit:
Make suggested changes

Co-authored-by: Fernando Rocha <34967844+Fernando-A-Rocha@users.noreply.github.com>

Dutchman101 committed:
Merge branch 'master' into serial
